crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,194 @@ use multimap::MultiMap;
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(name), span));
+                    }
+                    Some(..) => {
+                        self.stack.push((CheckInstructionKind::Replacing, span));
+                    }
+                    None => self.is_enabled = false,
+                }
+            }
+            CheckInstructionKind::Xoring => {
+                let right_branch = self.stack.pop();
+                let left_branch = self.stack.pop();
+                match (left_branch, right_branch) {
+                    (
+                        Some((CheckInstructionKind::PivotalNext(..), _)),
+                        Some((CheckInstructionKind::PivotalNext(..), _)),
+                    ) => {
+                        // `xor` has `next` in both branches. It is impossible
+                        // to tell which branch should poped up.
+                        self.stack.clear();
+                        self.is_enabled = false;
+                    }
+                    (Some((CheckInstructionKind::PivotalNext(iterator), _)), Some((..)))
+                    | (Some((..)), Some((CheckInstructionKind::PivotalNext(iterator), _))) => {
+                        // potential failure but need to check when fold pops up.
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(iterator), span));
+                    }
+                    (Some(_), Some(_)) => {
+                        self.stack.push((CheckInstructionKind::Xoring, span));
+                    }
+                    _ => {
+                        // disable machine if Xoring invariant, namely there must be 2 kinds on a stack, is broken.
+                        self.stack.clear();
+                        self.is_enabled = false;
+                    }
+                }
+            }
+            CheckInstructionKind::Merging => {
+                let right_branch = self.stack.pop();
+                let left_branch = self.stack.pop();
+                match (left_branch, right_branch) {
+                    (
+                        Some((CheckInstructionKind::PivotalNext(left_iterable), _)),
+                        Some((CheckInstructionKind::PivotalNext(right_iterable), _)),
+                    ) if left_iterable == right_iterable => {
+                        self.stack
+                            .push((CheckInstructionKind::PivotalNext(left_iterable), span));
+                    }
+                    (Some((CheckInstructionKind::PivotalNext(iterator), _)), Some((..))) => {
+                        // potential failure but need to check when fold pops up.
+                        self.stack.push((CheckInstructionKind::Merging, span));
+                        self.potentially_impossible_spans.push((iterator, span));
+                    }
+                    (Some((..)), Some((CheckInstructionKind::PivotalNext(iterable), _))) => {
```

Comment:
Why do you use both `..` and `_`?  Generally, `..` is used for "ignore rest of fields", not "all fields".

Reply:
Fixed